### ISSUE TITLE
what‘s the difference of performance and convergence between Adam and SGD optimizers with different shadular?

### ISSUE THREAD

**Issue by EricKani** [MIC-DKFZ/nnUNet#695]
what‘s the difference of performance and convergence between Adam and SGD optimizers with different shadular?
They are implemented in Trainer.py and TrainerV2.py respectively.

Thanks~
Eric

**Comment by FabianIsensee** [comment]
They are very similar but SGD was better in our experiments. If you plan to use Adam you need to adapt the learning rate to something lower as well.
Best,
Fabian

**Comment by EricKani** [comment]
Thanks~ I still use the version_1 nnUNet code by now [Facepalm]. Adam with 0.003 init_lr.

**Comment by EricKani** [comment]
In my experiments, I was really troubled in choicing a model to deploy when I use the "Adam" optimizer with "ReduceLROnPlateau" schedular...

When I use this training combination (adam + Plateau), the number of epoch in training phase is not definitive. And the "model_best.model" saved based on the moving average validation metric (Dice) is not truely the best. E.g., maybe I trained the model for 500 epoch in total, but the epoch of  model saved in  "model_best.model" is 250.

But when I use SGD with poly schedular, the number of training epochs is hard to decide. Maybe I have to run some experiments to determine the number of training epochs, because I Think the "model_final_checkpoint.model" is more reliable...

Are there some advices for choicing a combination from "Adam", "SGD","poly","plateau"? Thanks in advance!!

**Comment by FabianIsensee** [comment]
Honestly just let the regular nnUNet training run and pick model_final_checkpoint.model. Only consider using model_best.model when you see the green line in progress.ong to drop again. model_best.model uses an exponential moving average of the green line to determine what is considered 'best'. This is not very well tunes because I don't use it and it should also never be used to report 5-fold cross-validation scores (overfitting to validation set).
Please only modify the training length once you completed a full training and truly believe that shorter training could be beneficial. But it's probably even better to just modify the checkpointing behavior and save a separate checkpoint file every XXX epochs
Best,
Fabian
